gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroupStep.java

```diff
@@ -79,6 +79,14 @@ public void modulateBy(final Traversal.Admin<?, ?> kvTraversal) {
     }
 
     @Override
+    public void replaceLocalChild(final Traversal.Admin<?, ?> oldTraversal, final Traversal.Admin<?, ?> newTraversal) {
+        if (null != this.keyTraversal && this.keyTraversal.equals(oldTraversal))
+            this.keyTraversal = this.integrateChild(newTraversal);
+        else if (null != this.valueTraversal && this.valueTraversal.equals(oldTraversal))
```

Comment:
You will have to add all the `barrier` manipulation work in `replaceLocalChild` then. 

https://github.com/apache/tinkerpop/blob/185ad404011ebd1fb181c83b676ba8aa770c51c4/gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/GroupStep.java#L71-L74